### PR TITLE
feat: add `x/ecocredit` simulations

### PR DESCRIPTION
**Pull request by aleem1314**
## Description

Closes: #416 

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [x] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [x] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [x] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [x] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: x/ecocredit/exported/expected_keepers.go [hunk 1]
```diff
@@ -0,0 +1,25 @@
+package exported
```
Comment (ruhatch) [inline x/ecocredit/exported/expected_keepers.go:1]: @robert-zaremba you authored #353 what do you think we should do here?

File: x/ecocredit/simulation/genesis.go [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+
+	return supplies
+}
+
+// RandomizedGenState generates a random GenesisState for the ecocredit module.
+func RandomizedGenState(simState *module.SimulationState) {
+	startTime := simtypes.RandTimestamp(simState.Rand)
+	endTime := startTime.Add(24 * time.Hour)
+
+	//params
```
Comment (blushi) [inline x/ecocredit/simulation/genesis.go:149]:
```suggestion
	// params
```

File: x/ecocredit/simulation/operations.go [hunk 3]
```diff
@@ -0,0 +1,595 @@
...
+	}
+}
+
+// SimulateMsgCreateClass generates a MsgCreateClass with random values.
+func SimulateMsgCreateClass(ak exported.AccountKeeper, bk exported.BankKeeper,
+	qryClient ecocredit.QueryClient) simtypes.Operation {
+	return func(
+		r *rand.Rand, app *baseapp.BaseApp, sdkCtx sdk.Context, accs []simtypes.Account, chainID string,
+	) (simtypes.OperationMsg, []simtypes.FutureOperation, error) {
+		designer := accs[0]
```
Comment (blushi) [inline x/ecocredit/simulation/operations.go:121]: please update to use new terminology

File: x/ecocredit/simulation/operations.go [hunk 4]
```diff
@@ -0,0 +1,595 @@
...
+	issuers := make([]string, n)
+	for i := 0; i < n; i++ {
+		acc, _ := simtypes.RandomAcc(r, accounts)
+		issuers[i] = acc.Address.String()
+	}
+
+	return issuers
+}
+
+func generateBatchIssuence(r *rand.Rand, accs []simtypes.Account) []*ecocredit.MsgCreateBatch_BatchIssuance {
```
Comment (blushi) [inline x/ecocredit/simulation/operations.go:534]:
```suggestion
func generateBatchIssuance(r *rand.Rand, accs []simtypes.Account) []*ecocredit.MsgCreateBatch_BatchIssuance {
```

File: x/ecocredit/simulation/operations.go [hunk 5]
```diff
@@ -0,0 +1,595 @@
...
+			return simtypes.NoOpMsg(ecocredit.ModuleName, TypeMsgSend, err.Error()), nil, err
+		}
+
+		if tradableBalance.IsZero() || retiredBalance.IsZero() {
+			return simtypes.NoOpMsg(ecocredit.ModuleName, TypeMsgSend, "balance is zero"), nil, nil
+		}
+
+		recipient, _ := simtypes.RandomAcc(r, accs)
+		if batch.Issuer == recipient.Address.String() {
+			return simtypes.NoOpMsg(ecocredit.ModuleName, TypeMsgSend, "sender & reciever are same"), nil, nil
```
Comment (blushi) [inline x/ecocredit/simulation/operations.go:275]: let's use consistent terminology
```suggestion
			return simtypes.NoOpMsg(ecocredit.ModuleName, TypeMsgSend, "sender & recipient are same"), nil, nil
```

File: x/ecocredit/simulation/operations.go [hunk 6]
```diff
@@ -0,0 +1,595 @@
...
+			Bankkeeper:      bk,
+			ModuleName:      ecocredit.ModuleName,
+			CoinsSpentInMsg: spendable,
+		}
+
+		return GenAndDeliverTxWithRandFees(txCtx)
+	}
+}
+
+func getRandomClass(ctx regentypes.Context, r *rand.Rand, qryClient ecocredit.QueryClient,  msgType string) (*ecocredit.ClassInfo, simtypes.OperationMsg, error) {
```
Comment (blushi) [inline x/ecocredit/simulation/operations.go:484]:
```suggestion
func getRandomClass(ctx regentypes.Context, r *rand.Rand, qryClient ecocredit.QueryClient, msgType string) (*ecocredit.ClassInfo, simtypes.OperationMsg, error) {
```

File: x/ecocredit/simulation/operations.go [hunk 7]
```diff
@@ -0,0 +1,595 @@
...
+	if err != nil {
+		return nil, simtypes.NoOpMsg(ecocredit.ModuleName, msgType, err.Error()), err
+	}
+
+	classes := res.Classes
+	if len(classes) == 0 {
+		return nil, simtypes.NoOpMsg(ecocredit.ModuleName, msgType, "no credit class found"), nil
+	}
+
+	return classes[r.Intn(len(classes))], simtypes.NoOpMsg(ecocredit.ModuleName, msgType,""),nil
```
Comment (blushi) [inline x/ecocredit/simulation/operations.go:495]:
```suggestion
	return classes[r.Intn(len(classes))], simtypes.NoOpMsg(ecocredit.ModuleName, msgType, ""),nil
```

File: x/ecocredit/simulation/operations.go [hunk 8]
```diff
@@ -0,0 +1,595 @@
...
+	if err != nil {
+		return nil, simtypes.NoOpMsg(ecocredit.ModuleName, msgType, err.Error()), err
+	}
+
+	batches := res.Batches
+	if len(batches) == 0 {
+		return nil, simtypes.NoOpMsg(ecocredit.ModuleName, msgType, "no batch found"), nil
+	}
+
+	return batches[r.Intn(len(batches))], simtypes.NoOpMsg(ecocredit.ModuleName, msgType,""),nil
```
Comment (blushi) [inline x/ecocredit/simulation/operations.go:511]:
```suggestion
	return batches[r.Intn(len(batches))], simtypes.NoOpMsg(ecocredit.ModuleName, msgType, ""),nil
```

File: x/ecocredit/simulation/operations.go [hunk 9]
```diff
@@ -0,0 +1,595 @@
...
+
+	classes := res.Classes
+	if len(classes) == 0 {
+		return nil, simtypes.NoOpMsg(ecocredit.ModuleName, msgType, "no credit class found"), nil
+	}
+
+	return classes[r.Intn(len(classes))], simtypes.NoOpMsg(ecocredit.ModuleName, msgType,""),nil
+}
+
+func getRandomBatchFromClass(ctx regentypes.Context, r *rand.Rand, qryClient ecocredit.QueryClient,  msgType, classID string) (*ecocredit.BatchInfo, simtypes.OperationMsg, error) {
```
Comment (blushi) [inline x/ecocredit/simulation/operations.go:498]:
```suggestion
func getRandomBatchFromClass(ctx regentypes.Context, r *rand.Rand, qryClient ecocredit.QueryClient, msgType, classID string) (*ecocredit.BatchInfo, simtypes.OperationMsg, error) {
```

**Review by blushi — COMMENTED** [review]
Looks good, except some nits/ language updates / lint issues

File: x/ecocredit/simulation/operations.go [hunk 10]
```diff
@@ -0,0 +1,594 @@
...
+	OpWeightMsgCreateBatch = "op_weight_msg_create_batch"
+	OpWeightMsgSend        = "op_weight_msg_send"
+	OpWeightMsgRetire      = "op_weight_msg_retire"
+	OpWeightMsgCancel      = "op_weight_msg_cancel"
+)
+
+// ecocredit operations weights
+const (
+	WeightCreateClass = 100
+	WeightCreateBatch = 80
```
Comment (ruhatch) [inline x/ecocredit/simulation/operations.go:32]: I think we want to generate a lot more classes than batches, so maybe this weight should be higher?

File: x/ecocredit/simulation/operations.go [hunk 11]
```diff
@@ -0,0 +1,594 @@
...
+	WeightSend        = 70
```
Comment (ruhatch) [inline x/ecocredit/simulation/operations.go:33]: Also I think we want more sends than anything else, then probably more retires then creation messages.

File: x/ecocredit/simulation/operations.go [hunk 12]
```diff
@@ -0,0 +1,594 @@
...
+	WeightRetire      = 70
+	WeightCancel      = 70
```
Comment (ruhatch) [inline x/ecocredit/simulation/operations.go:35]: Maybe more like this, if I've understood the weighting correctly...
```suggestion
	WeightCreateClass = 10
	WeightCreateBatch = 50
	WeightSend        = 100
	WeightRetire      = 80
	WeightCancel      = 30
```

File: x/ecocredit/simulation/operations.go [hunk 13]
```diff
@@ -0,0 +1,594 @@
...
+			return simtypes.NoOpMsg(ecocredit.ModuleName, TypeMsgCreateClass, err.Error()), nil, err
+		}
+
+		params := res.Params
+		if params.AllowlistEnabled && !contains(params.AllowedClassCreators, admin.Address.String()) {
+			return simtypes.NoOpMsg(ecocredit.ModuleName, TypeMsgCreateClass, "not allowed to create credit class"), nil, nil // skip
+		}
+
+		spendable := bk.SpendableCoins(sdkCtx, admin.Address)
+		if spendable.IsAllLTE(params.CreditClassFee) {
```
Comment (ruhatch) [inline x/ecocredit/simulation/operations.go:135]: Are we sure we're generating cases where this is false, i.e. the admin has enough funds?

File: x/ecocredit/simulation/operations.go [hunk 14]
```diff
@@ -0,0 +1,594 @@
...
+	for i := 0; i < n; i++ {
+		acc, _ := simtypes.RandomAcc(r, accounts)
+		issuers[i] = acc.Address.String()
+	}
+
+	return issuers
+}
+
+func generateBatchIssuance(r *rand.Rand, accs []simtypes.Account) []*ecocredit.MsgCreateBatch_BatchIssuance {
+	numIssuences := simtypes.RandIntBetween(r, 3, 10)
```
Comment (ruhatch) [inline x/ecocredit/simulation/operations.go:534]:
```suggestion
	numIssuances := simtypes.RandIntBetween(r, 3, 10)
```

File: x/ecocredit/simulation/operations.go [hunk 15]
```diff
@@ -0,0 +1,594 @@
...
+	res := make([]*ecocredit.MsgCreateBatch_BatchIssuance, numIssuences)
+
+	for i := 0; i < numIssuences; i++ {
+		recipient := accs[i]
+		res[i] = &ecocredit.MsgCreateBatch_BatchIssuance{
+			Recipient:          recipient.Address.String(),
+			TradableAmount:     "12345.123123",
```
Comment (ruhatch) [inline x/ecocredit/simulation/operations.go:541]: Maybe randomise these?

File: x/ecocredit/simulation/operations.go (hunk 10 above)
Comment (aleem1314) [reply x/ecocredit/simulation/operations.go:32]: Yes, we are generating more classes than batches

```
"ecocredit": {
  "/regen.ecocredit.v1alpha1.MsgCreateBatch": {
   "ok": 73,
   "failure": 0
  },
  "/regen.ecocredit.v1alpha1.MsgCreateClass": {
   "ok": 118,
   "failure": 0
  }
 }
 ```

File: x/ecocredit/simulation/operations.go (hunk 13 above)
Comment (aleem1314) [reply x/ecocredit/simulation/operations.go]: Yes. Without this check, simulations are failing.

```
failed to execute message; message index: 0: 4stake is smaller than 32stake: insufficient funds [regen-network/cosmos-sdk@v0.43.0-regen-1/x/bank/keeper/send.go:187]
```

File: x/ecocredit/simulation/genesis.go [hunk 16]
```diff
@@ -0,0 +1,242 @@
...
+	supply               = "supplies"
+	classFee             = "credit_class_fee"
+	allowedCreators      = "allowed_class_creators"
+	typeAllowListEnabled = "allow_list_enabled"
+	typeCreditTypes      = "credit_types"
+)
+
+// genCreditClassFee randomized CreditClassFee
+func genCreditClassFee(r *rand.Rand) sdk.Coins {
+	return sdk.NewCoins(sdk.NewCoin(sdk.DefaultBondDenom, sdk.NewInt(1)))
```
[the remark this answers is not in the document]
Comment (aleem1314) [reply x/ecocredit/simulation/genesis.go]: updated -> [e4c8439](https://github.com/regen-network/regen-ledger/pull/480/files#diff-e4c8439f874c9a734d9cf5c8d205173f102c6d20fc7bf314e54ecdb41c8b071cR32)

File: types/module/server/manager.go [hunk 17]
```diff
@@ -80,7 +80,7 @@ func (mm *Manager) RegisterGRPCGatewayRoutes(apiSvr *api.Server) {
 
 // RegisterModules registers modules with the Manager and registers their services.
 func (mm *Manager) RegisterModules(modules []module.Module) error {
-	mm.modules = modules
+	mm.modules = append(mm.modules, modules...)
```
Comment (anilCSE) [inline types/module/server/manager.go:83]: we can avoid this by moving ecocredit module wiring into stable config.

File: types/module/server/manager.go [hunk 18]
```diff
@@ -145,7 +145,7 @@ func (mm *Manager) RegisterModules(modules []module.Module) error {
 		mm.exportGenesisHandlers[name] = cfg.exportGenesisHandler
 
 		if cfg.weightedOperationHandler != nil {
-			mm.weightedOperationsHandlers[name] = cfg.weightedOperationHandler
+			mm.weightedOperationsHandlers = append(mm.weightedOperationsHandlers, cfg.weightedOperationHandler)
```
Comment (anilCSE) [inline types/module/server/manager.go:148]: same as above

File: types/module/server/manager.go (hunk 17 above)
Comment (anilCSE) [reply types/module/server/manager.go:83]: update, PR here: https://github.com/regen-network/regen-ledger/pull/524

File: x/ecocredit/simulation/genesis.go [hunk 19]
```diff
@@ -0,0 +1,242 @@
...
+		{
+			Name:         "biodiversity",
+			Abbreviation: "BIO",
+			Unit:         "ton",
+			Precision:    6,
+		},
+	}
+}
+
+func genClasses(r *rand.Rand, accounts []simtypes.Account) []*ecocredit.ClassInfo {
```
[the remark this answers is not in the document]
Comment (aleem1314) [reply x/ecocredit/simulation/genesis.go]: > Should we be using genCreditTypes in this function then randomly selecting one of the two types generated by that?

Yeah, It makes sense to randomize `creditTypes`.

File: types/module/server/manager.go (hunk 17 above)
Comment (anilCSE) [reply types/module/server/manager.go:83]: ah, that would duplicate the integration code for ecocredit module in stable and experimental modules. Let's keep it as-is.

File: x/ecocredit/simulation/params.go [hunk 20]
```diff
@@ -0,0 +1,69 @@
...
+
+// ParamChanges defines the parameters that can be modified by param change proposals
+// on the simulation
+func ParamChanges(r *rand.Rand) []simtypes.ParamChange {
+	allowListEnabled := false
+
+	return []simtypes.ParamChange{
+		simulation.NewSimParamChange(ecocredit.ModuleName, string(ecocredit.KeyCreditClassFee),
+			func(r *rand.Rand) string {
+				bz, err := json.Marshal(genCreditClassFee(r))
```
[the remark this answers is not in the document]
Comment (aleem1314) [reply x/ecocredit/simulation/params.go:22]: I think No. We don't have a reference to the app codec.

File: x/ecocredit/simulation/genesis_test.go [hunk 21]
```diff
@@ -0,0 +1,59 @@
...
+		InitialStake: 1000,
+		GenState:     make(map[string]json.RawMessage),
+	}
+
+	simulation.RandomizedGenState(&simState)
+
+	var ecocreditGenesis ecocredit.GenesisState
+	simState.Cdc.MustUnmarshalJSON(simState.GenState[ecocredit.ModuleName], &ecocreditGenesis)
+
+	require.Equal(t, ecocreditGenesis.Params.AllowedClassCreators, []string{"cosmos1tnh2q55v8wyygtt9srz5safamzdengsnqeycj3"})
```
[the remark this answers is not in the document]
Comment (aleem1314) [reply x/ecocredit/simulation/genesis_test.go:43]: I don't think we can check invariants here,  they are private functions in `server`.

**Review by anilCSE — APPROVED** [review]
lgtm

**Review by ruhatch — APPROVED** [review]
LGTM! Thanks :)

**Pull request merged**